### PR TITLE
[quorum store][mempool] optimize pulling txns for when many txns are in progress

### PR DESCRIPTION
**Pull request by bchocho**
### Description

In forge `realistic_network_tuned_for_throughput` it was found that:
* Calling pull takes > 25 ms.
* In an edge case where the pull returns 0 new txns, this could cause livelock because the pull call would be called every 25 ms.

Resolved by:
* Optimizing the pull; now on the same workload the pull takes ~8 ms.
  * Instead of using a hashmap, use a sorted vector + binary search for the excluded txns
  * And only re-sort the vector when batches in progress is updated (instead of every pull)
* Checking the time taken to pull, and pulling less frequently if the pull takes > 12.5 ms.

### Test Plan

Existing tests, ran `realistic_network_tuned_for_throughput` as ad-hoc job and observed results.

### CODE REVIEW

**Comment by bchocho** [comment]
Recent ad-hoc test: https://github.com/aptos-labs/aptos-core/actions/runs/6725257657/attempts/1#summary-18279247367
vs forge stable: https://github.com/aptos-labs/aptos-core/actions/runs/6716280366/attempts/1#summary-18257624474

**Review by zekun000 — COMMENTED** [review]
should we remove the biased in the select? or moving the command branch up?

**Comment by bchocho** [comment]
Good point. I'll move the command branch up, since that can cause problems upstream.

File: consensus/src/quorum_store/batch_generator.rs [hunk 1]
```diff
@@ -292,17 +311,72 @@ impl BatchGenerator {
...
+                        }
+                        BatchGeneratorCommand::Shutdown(ack_tx) => {
+                            ack_tx
+                                .send(())
+                                .expect("Failed to send shutdown ack");
+                            break;
+                        },
+                    }
+                }),
                 Some(updated_back_pressure) = back_pressure_rx.recv() => {
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_generator.rs:369]: I think we want to process back pressure before the cmd?

**Comment by zekun000** [comment]
also I don't remember why this select requires a biased, if we don't have strong reasons we should default to use the randomized select

**Comment by bchocho** [comment]
I vaguely remember in a previous version, the backpressure was on/off and so having it absolutely up-to-date was a concern. However, with the current backpressure it's probably less of a concern. I'll remove biased and run some experiments

**Comment by bchocho** [comment]
Things look fine without biased. Here is a run with gauge instrumented (from #10772), it has at max 1 item in the queue: https://github.com/aptos-labs/aptos-core/actions/runs/6736434894

**Review by zekun000 — COMMENTED** [review]
from what I understand, it feels a BTreeMap is better than a Vec for the exclude_transactions_sorted?

**Comment by bchocho** [comment]
For BTreeMap, it's hard to know if it's safe to remove a transaction from exclude_transactions_sorted when a batch is removed from batches_in_progress. I'm pretty sure there are edge cases when there could be transactions in multiple batches_in_progress, so we couldn't remove _all_ transactions within the removed batch. I suppose we could do some kind of reference counting or check the batch id before removal, but that makes it somewhat complex.

With the current approach, exclude_transactions_sorted is computed from scratch, so it's simpler (but possibly more inefficient)

File: consensus/src/quorum_store/batch_generator.rs [hunk 2]
```diff
@@ -217,22 +219,65 @@ impl BatchGenerator {
...
+        }
+        self.batches_in_progress.insert(batch_id, txns);
+    }
+
+    fn remove_batch_in_progress(&mut self, batch_id: &BatchId) -> bool {
+        let removed = self.batches_in_progress.remove(batch_id);
+        match removed {
+            Some(txns) => {
+                for txn in txns {
+                    if let Some(mut info) = self.txns_in_progress_sorted.remove(&txn) {
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_generator.rs:245]: nit: this looks weird, I thought the idiomatic way is to get an entry and delete the entry if decrement to 0

File: consensus/src/quorum_store/batch_generator.rs [hunk 3]
```diff
@@ -217,22 +219,65 @@ impl BatchGenerator {
...
+        txns_in_progress: Vec<(TransactionSummary, TransactionInProgress)>,
+    ) {
+        let mut txns = vec![];
+        for (summary, info) in txns_in_progress {
+            let txn_info = self
+                .txns_in_progress_sorted
+                .entry(summary)
+                .or_insert_with(|| TransactionInProgress::new(info.gas_unit_price));
+            txn_info.increment();
+            txn_info.gas_unit_price = info.gas_unit_price.max(txn_info.gas_unit_price);
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_generator.rs:234]: should we overwrite with higher gas price? i.e. don't increment and decrement if gas unit price not equal

File: mempool/src/core_mempool/mempool.rs [hunk 4]
```diff
@@ -234,30 +234,24 @@ impl Mempool {
...
+                        upgraded.insert(txn_pointer);
                     }
                 }
             }
-            for txn_pointer in seen_and_upgraded {
-                seen.remove(txn_pointer);
-            }
         }
+        let gas_end_time = start_time.elapsed();
+        let gas_time = gas_end_time;
```
Comment (zekun000) [inline mempool/src/core_mempool/mempool.rs:254]: not sure i understand why we need two variables here

File: consensus/src/quorum_store/batch_generator.rs [hunk 5]
```diff
@@ -217,22 +219,65 @@ impl BatchGenerator {
...
+        );
 
         let mut pulled_txns = self
             .mempool_proxy
             .pull_internal(
                 max_count,
                 self.config.mempool_txn_pull_max_bytes,
-                exclude_txns,
+                // TODO: pass the actual BTreeMap
+                self.txns_in_progress_sorted.clone(),
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_generator.rs:280]: feels like Arc of mutex can avoid the clone, not sure how much it matters.

**Review by zekun000 — APPROVED** [review]

**Review by JoshLind — APPROVED** [review]

**Pull request merged**